src/background/StreamSessionManager.ts

```diff
@@ -0,0 +1,373 @@
+                    // If the current user is the new leader, they should do a full sync after a processing wait
+                    setTimeout(() => {
+                        const syncStateMessage = this.generateSyncStateMessageFromCurrentState();
+                        this.sessionState.sessionUsers.forEach(user => {
+                            if (user.id !== this.currentUserId) {
+                                user.connection?.send(syncStateMessage);
+                            }
+                        });
+                    }, 500);
+                }
+            }
+        });
+    }
+
+    /**
+     * Determines the new leader id after the leader disconnects. Assumes the previous leader has been removed from sessionState
+     */
+    private determineNewLeaderId(): string {
+        for (const userId of this.sessionState.sessionJoinOrder) {
+            const user = this.sessionState.sessionUsers.get(userId);
+            if (user?.permissions.canControlVideo) {
+                return user.id;
+            }
+        }
+
+        // If we haven't found any other users who can control video, default to the oldest in the join order
+        return this.sessionState.sessionJoinOrder[0];
+    }
+
+    /**
+     * Handler for state sync request messages
+     */
+    private handleStateSyncRequestMessage(): void {
+        console.warn("state sync requested");
+
+        if (this.sessionState.leaderId !== this.currentUserId) {
+            console.warn(`Received sync state request on ${this.currentUserId} but leader is: ${this.sessionState.leaderId}`);
+            return;
+        }
+
+        const stateSyncMessage = this.generateSyncStateMessageFromCurrentState();
+        this.sessionState.sessionUsers.forEach(user => {
+            if (user.id !== this.currentUserId) {
+                console.warn(`Sending sync state to: ${user.id}`);
+                user.connection?.send(stateSyncMessage);
+            }
+        });
+    }
+
+    /**
+     * Handler for state sync messages
+     * @param message The state sync message to process
+     */
+    private handleStateSyncMessage(message: IStateSyncMessage): void {
+        console.warn("handling state sync");
+        const numUsersBefore = this.sessionState.sessionUsers.size;
+        // List of ids used for cleaning up users list after updates
+        const newListIds: string[] = [];
+
+        for (const user of message.state.sessionUsers) {
+            const existingUser = this.sessionState.sessionUsers.get(user.id);
+            const existingUserPeerJsConnection = existingUser?.connection;
+
+            const newUserObj: IUserConnection = {
+                id: user.id,
+                displayName: user.displayName,
+                connection: existingUserPeerJsConnection,
+                permissions: user.permissions
+            };
+            this.sessionState.sessionUsers.set(user.id, newUserObj);
+            newListIds.push(user.id);
```

Comment:
Looking at `newUsers` more closely, I'm also wondering if it's necessary to even do a `set()` with an undefined `connection` value. Since it's just going to get overwritten in a few ms anyway, when they actually do connect.

Reply:
I like the idea of having the state set to exactly what we know it to be instead of a special case. I also think undefined connection value helps draw attention to the async nature of setting up the connections. Also if you're bringing this up to avoid allowing undefined in the IUserConnection interface, since we have the currentUser there we'd still need it to be undefined in that case